### PR TITLE
Reduce contention a little in AbstractPartitionedLimiter

### PR DESCRIPTION
**Pull request by jasonk000**
Swap `busy` from an `int` to `AtomicInteger`, which eliminates need for a `release` to take the lock.

And while we are here, slim down the work done while holding the lock.

### CODE REVIEW

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java [hunk 1]
```diff
@@ -216,6 +221,7 @@ public Optional<Listener> acquire(ContextT context) {
         try {
             lock.lock();
             if (shouldBypass(context)){
+                lock.unlock();
```
Comment (enbnt) [inline concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java:224]: Do we even need to lock in the `shouldBypass` scenario?

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java [hunk 2]
```diff
@@ -105,10 +105,10 @@ public Limiter<ContextT> build() {
 
     static class Partition {
         private final String name;
+        private final AtomicInteger busy = new AtomicInteger(0);
 
         private double percent = 0.0;
         private int limit = 0;
```
Comment (enbnt) [inline concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java:111]: Calls to `updateLimit` don't appear to guarantee thread-safety. Any reason we shouldn't move this to `AtomicInteger` as well?

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java [hunk 3]
```diff
@@ -134,25 +134,30 @@ void updateLimit(int totalLimit) {
...
 
         boolean isLimitExceeded() {
-            return busy >= limit;
+            return busy.get() >= limit;
         }
 
-        void acquire() {
-            busy++;
-            inflightDistribution.addSample(busy);
+        int acquire() {
```
Comment (enbnt) [inline concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java:140]: this is technically a binary incompatible breaking change. Can we instead introduce this as a new `private final int acquireAndGet()` method and `private final void acquirePostSample(int nowBusy)`, while `@Deprecated` the existing `void acquire()` which calls into those methods?

It looks like the big change here is the move to using atomic and splitting up the behavior in the public `acquire` method to more narrowly scope the lock. It's possible I'm missing something to where the new methods couldn't be privately scoped or if the default scoped `acquire` method should change its behavior, but it's not obvious from the change alone.

Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java:140]: This is a great solution, thanks!

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java (hunk 1 above)
Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java:224]: The issue I was guarding for is that it's not enforced. That [could be overriden by any implementation](https://github.com/Netflix/concurrency-limits/blob/075bc76be8b90172150699dbb2f28c6f85c71e97/concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractLimiter.java#L116-L118).

I think if we got shouldBypass removed from the lock, then the only reason we need the lock is for getInflight >= getLimit check, and I _think_ we can do that lock-free as well.

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java (hunk 2 above)
Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java:111]: Can do that, no reason not to.

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java (hunk 1 above)
Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java:246]: https://github.com/Netflix/concurrency-limits/blob/075bc76be8b90172150699dbb2f28c6f85c71e97/concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractLimiter.java#L73-L74

Would it be a viable solution to expand the scope of this comment to indicate that the predicate/subclasses should not query the limiter, and if they do so they are responsible for concurrency?

About the only situation I can see this happening is where the shouldBypass might make a dynamic decision based on the `getInflight` value. I'm not sure what real-world configurations exist.

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java (hunk 3 above)
Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java]: fixed in https://github.com/Netflix/concurrency-limits/pull/204/commits/794dac3d35ccf6f1ca586e509c0f3f9164750b75

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java (hunk 1 above)
Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java:246]: Looking over at these examples -- [GrpcServerLimiterBuilder.java](https://github.com/Netflix/concurrency-limits/blob/075bc76be8b90172150699dbb2f28c6f85c71e97/concurrency-limits-grpc/src/main/java/com/netflix/concurrency/limits/grpc/server/GrpcServerLimiterBuilder.java#L49-L93) and [ServletLimiterBuilder.java](https://github.com/Netflix/concurrency-limits/blob/075bc76be8b90172150699dbb2f28c6f85c71e97/concurrency-limits-servlet/src/main/java/com/netflix/concurrency/limits/servlet/ServletLimiterBuilder.java#L77-L143) it seems that none of them are built to refer back to the limiter itself. If this is the expected & existing usage, it seems sensible to me that we could move the bypass check outside the lock and provide a comment explicitly.

Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java]: I've gone with this solution which has allowed for a non-blocking normal case solution. This also allowed to roll back some of the other acquire and deprecated changes, etc. See latest!

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java [hunk 4]
```diff
@@ -213,61 +210,60 @@ private Partition resolvePartition(ContextT context) {
...
-                }
+            return createRejectedListener();
+        }
 
-                @Override
-                public void onIgnore() {
-                    listener.onIgnore();
-                    releasePartition(partition);
-                }
+        partition.acquire();
```
Comment (jasonk000) [inline concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java:241]: let me noodle on this some more, it should probably be more atomic

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java (hunk 1 above)
Comment (enbnt) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java]: Just chiming in to say that the changes are taking shape nicely, let us know when you're done noodling 😸

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java (hunk 4 above)
Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java]: fixed in 7dd2ae8

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java (hunk 1 above)
Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java]: I think this is ready for another review.

🙏 thank you for patience.

File: concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java [hunk 5]
```diff
@@ -134,25 +133,41 @@ void updateLimit(int totalLimit) {
...
+            while (true) {
+                int current = busy.get();
+                if (current >= limit) {
+                    return false;
+                }
+                if (busy.compareAndSet(current, current + 1)) {
+                    inflightDistribution.addSample(current + 1);
+                    return true;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (jasonk000) [reply concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiter.java]: fixed in 06322db

File: concurrency-limits-core/src/test/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiterTest.java [hunk 6]
```diff
@@ -227,4 +236,86 @@ public void testBypassSimpleLimiter() {
...
+
+        startLatch.countDown();
+        endLatch.await();
+        executor.shutdown();
+        executor.awaitTermination(10, TimeUnit.SECONDS);
+
+        for (String partition : Arrays.asList("A", "B", "C")) {
+            System.out.println(partition + " success count: " + successCounts.get(partition).get());
+            System.out.println(partition + " rejection count: " + rejectionCounts.get(partition).get());
+            System.out.println(partition + " max concurrent: " + maxConcurrents.get(partition).get());
```
Comment (enbnt) [inline concurrency-limits-core/src/test/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiterTest.java:304]: nit: can we avoid System.out.println's in the tests? I think it's ok to build an append this string to the assertions if they fail as a hint.

File: concurrency-limits-core/src/test/java/com/netflix/concurrency/limits/limiter/SimpleLimiterTest.java [hunk 7]
```diff
@@ -82,4 +90,63 @@ public void testSimpleBypassLimiterDefault() {
...
+        }
+
+        startLatch.countDown();
+        endLatch.await();
+        executor.shutdown();
+        executor.awaitTermination(10, TimeUnit.SECONDS);
+
+        System.out.println("Success count: " + successCount.get());
+        System.out.println("Rejection count: " + rejectionCount.get());
+        System.out.println("Max concurrent: " + maxConcurrent.get());
```
Comment (enbnt) [inline concurrency-limits-core/src/test/java/com/netflix/concurrency/limits/limiter/SimpleLimiterTest.java:145]: same re: avoiding System.out in tests

**Review by fedorka — APPROVED** [review]

**Review by enbnt — APPROVED** [review]

File: concurrency-limits-core/src/test/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiterTest.java (hunk 6 above)
Comment (jasonk000) [reply concurrency-limits-core/src/test/java/com/netflix/concurrency/limits/limiter/AbstractPartitionedLimiterTest.java]: fixed in [09a1e6f](https://github.com/Netflix/concurrency-limits/pull/204/commits/09a1e6fd1d649b881c067d18f294011f2360cbe5)

File: concurrency-limits-core/src/test/java/com/netflix/concurrency/limits/limiter/SimpleLimiterTest.java (hunk 7 above)
Comment (jasonk000) [reply concurrency-limits-core/src/test/java/com/netflix/concurrency/limits/limiter/SimpleLimiterTest.java]: fixed in [09a1e6f](https://github.com/Netflix/concurrency-limits/pull/204/commits/09a1e6fd1d649b881c067d18f294011f2360cbe5)

**Review by umairk79 — APPROVED** [review]

**Pull request merged**